lib/ddtrace/tracing/runtime.rb

```diff
@@ -0,0 +1,90 @@
+require 'forwardable'
+require 'ddtrace/configuration/settings'
+require 'ddtrace/configuration/components'
+require 'ddtrace/configuration/pin_setup'
+
+module Datadog
+  module Tracing
+    # This class encapsulates all tracer runtime data and objects.
+    #
+    # Any stateful component or information that the tracer requires
+    # to execute must be directly or indirectly managed by {Runtime}.
+    #
+    # Any object who's life cycle is not captured by {Runtime} should be
+    # part of the generic tracer structure, unaffected by configuration
+    # or runtime state.
+    #
+    # Creating a new {Runtime} will effectively configure and initialized a new
+    # running instance of the tracer.
+    #
+    # Reconfiguration the {Runtime} will affect the running tracer instance without
+    # a complete restart of the tracer internals.
+    #
+    # Destroying the {Runtime} will shutdown the current tracer instance.
+    #
+    # The tracer has started if and only if (iff) the {Runtime} has been started.
+    # The tracer is shut down if and only if (iff) the {Runtime} is shut down.
+    #
+    # A production user of the tracer will only initialize at most a single {Runtime}
+    # instance throughout the lifetime of the host application. Changes to the tracer
+    # state will be handled by {Runtime#configure} invocations.
+    #
+    # TODO: {Runtime} currently indirectly manages the Profiler, though Components.
+    # TODO: The Profiler life cycle management should be eventually extracted to its own runtime.
+    class Runtime
```

Comment:
Should this note extend to a some of the items below as well? logger and runtime metrics shouldn't (imho) be specific to tracing either.

Reply:
Agree, there should be a `Runtime` per product, handling its own things. And I guess one for shared stuff like the logger.